improve CRD handling in VolumePopulator test

RomanBednar opened the pull request:
#### What type of PR is this?
/kind bug
/kind failing-test
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Fixes e2e test, see commit message for details.

#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->

#### Special notes for your reviewer:

Patch verified with the steps below:

#### 1) Run test with CRD installed in cluster

Create VolumePopulator CRD:
```
$ oc create -f /Users/rbednar/git/origin/vendor/k8s.io/kubernetes/test/e2e/testing-manifests/storage-csi/any-volume-datasource/crd/populator.storage.k8s.io_volumepopulators.yaml
customresourcedefinition.apiextensions.k8s.io/volumepopulators.populator.storage.k8s.io created
```

Run test, check CRD is detected and left unchanged:
```
I0722 11:12:31.447465 41055 namespace.go:59] About to run a Kube e2e test, ensuring namespace/e2e-provisioning-1979-val-6477 is privileged
    STEP: Checking if VolumePopulator CRD volumepopulators.populator.storage.k8s.io exists @ 07/22/25 11:12:32.576
    STEP: VolumePopulator CRD volumepopulators.populator.storage.k8s.io already exists, leaving unchanged @ 07/22/25 11:12:32.69
    STEP: Deploying validator @ 07/22/25 11:12:32.69
.
.
.
  ------------------------------

  Ran 1 of 1 Specs in 84.469 seconds
  SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
```


#### 2) Run test without CRD installed in cluster

Remove CRD:
```
$ oc delete crd/volumepopulators.populator.storage.k8s.io
customresourcedefinition.apiextensions.k8s.io "volumepopulators.populator.storage.k8s.io" deleted
```

Run test, check CRD is created and cleaned up:
```
I0722 11:17:29.329789 41190 namespace.go:59] About to run a Kube e2e test, ensuring namespace/e2e-provisioning-8379-val-3961 is privileged
    STEP: Checking if VolumePopulator CRD volumepopulators.populator.storage.k8s.io exists @ 07/22/25 11:17:30.561
    STEP: VolumePopulator CRD not found, test will create it and remove when done @ 07/22/25 11:17:30.761
    STEP: Deploying validator @ 07/22/25 11:17:30.761
    STEP: Creating VolumePopulator CRD @ 07/22/25 11:17:30.761
I0722 11:17:30.763400 41190 create.go:157] creating *v1.CustomResourceDefinition: volumepopulators.populator.storage.k8s.io
.
.
.
    STEP: Cleaning up VolumePopulator CRD: volumepopulators.populator.storage.k8s.io @ 07/22/25 11:18:53.603
.
.
.
  ------------------------------

  Ran 1 of 1 Specs in 96.070 seconds
  SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
```

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

dobsonj commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -297,12 +299,73 @@ func (p *provisioningTestSuite) DefineTests(driver storageframework.TestDriver,
+				crdManifestPath)
+			framework.ExpectNoError(err)
+
+			ginkgo.DeferCleanup(func(ctx context.Context) {
+				ginkgo.By(fmt.Sprintf("Cleaning up %s CRD", crd.Name))
+				err := apiExtensionClient.ApiextensionsV1().CustomResourceDefinitions().Delete(ctx, crd.Name, metav1.DeleteOptions{})
+				if err != nil && !apierrors.IsNotFound(err) {
+					framework.Logf("Failed to delete %s CRD: %v", crd.Name, err)
+				}
+			})
```
I would probably move this `DeferCleanup` block directly into the "CRD creation succeeded" block instead of using the `crdCreatedByTest` boolean.

RomanBednar replied:
Removed - applied suggestion with using cleanup from `storageutils.CreateFromManifests`.

dobsonj commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -297,12 +299,73 @@ func (p *provisioningTestSuite) DefineTests(driver storageframework.TestDriver,
+				}
+			})
+
+			ginkgo.By(fmt.Sprintf("Waiting for %s CRD to be created", crd.Name))
+
+			clusterCrds := framework.RetryNotFound(framework.HandleRetry(func(ctx context.Context) (*metav1.APIResourceList, error) {
+				return apiExtensionClient.Discovery().ServerResourcesForGroupVersion("populator.storage.k8s.io/v1beta1")
+			}))
+			includePopulatorCrd := framework.MakeMatcher(func(actual *metav1.APIResourceList) (func() string, error) {
+				for _, resource := range actual.APIResources {
+					if resource.Kind == crd.Kind {
+						return nil, nil
+					}
+				}
+				return func() string {
+					return "VolumePopulator CRD not found"
+				}, nil
+			})
+			err = framework.Gomega().Eventually(ctx, clusterCrds).Should(includePopulatorCrd)
+			framework.ExpectNoError(err)
```
Why the additional retry logic? Is it necessary for the CRD? If so, why is it different from the other manifests?

dobsonj commented:
/triage accepted
/priority important-soon

mkimuram commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -297,12 +299,73 @@ func (p *provisioningTestSuite) DefineTests(driver storageframework.TestDriver,
+
+		var crdCreatedByTest bool
+
+		ginkgo.By(fmt.Sprintf("Checking if %s CRD exists", crd.Name))
+		_, err = apiExtensionClient.ApiextensionsV1().CustomResourceDefinitions().Get(ctx, crd.Name, metav1.GetOptions{})
+		if err != nil && apierrors.IsNotFound(err) {
+			ginkgo.By("VolumePopulator CRD not found, test will create it and remove when done")
+			crdCreatedByTest = true
+		} else if err != nil {
+			framework.ExpectNoError(err, "Error checking for VolumePopulator CRD existence")
+		} else {
+			ginkgo.By(fmt.Sprintf("%s CRD already exists, leaving unchanged", crd.Name))
+			crdCreatedByTest = false
+		}
```
Another idea would be to add "populator.storage.k8s.io_volumepopulators.yaml" to valManifests if  the VolumePopulator CRD doesn't already exsit. Then, `storageutils.CreateFromManifests` will handle cleanup for both case properly without additional logics.

RomanBednar replied:
I like what @mkimuram suggested - we can keep the check and add the CRD manifest conditionally to achieve the same thing with less code. Thanks!

RomanBednar committed:
improve CRD handling in VolumePopulator test

Test: provisioning should provision storage with any volume data source

During CSI certification test we observed that the test can fail
with a message:
"customresourcedefinitions.apiextensions.k8s.io \"volumepopulators.populator.storage.k8s.io\" already exists"

This is because the test does not consider that this CRD can be already
installed in the cluster.

The test was updated to handle the CRD better by creating it for the
duration of the test and removing it afterward. Otherwise, if the CRD
is already installed, the test will neither create nor remove it

dobsonj commented:
Thanks @RomanBednar , there is one linter warning, but aside from that the latest changes look good :+1:

RomanBednar commented:
/retest-required

RomanBednar commented:
/retest-required

RomanBednar commented:
/milestone v1.34

dobsonj commented:
/lgtm

RomanBednar commented:
/assign @gnufied 

For approval.

gnufied commented:
/approve
